File: app/src/main/java/prototypez/github/io/sq/demo/fragment/LoginByPwdFragment.java
package prototypez.github.io.sq.demo.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import java.util.UUID;

import prototypez.github.io.sq.demo.R;
import prototypez.github.io.sq.demo.entity.User;

public class LoginByPwdFragment extends Fragment {


    // UI references.
    private EditText mPhoneView;
    private EditText mPasswordView;

    LoginByPwdCallback mCallback;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login_by_pwd, container, false);
        // Set up the login form.
        mPhoneView = view.findViewById(R.id.phone);

        mPasswordView = view.findViewById(R.id.password);

        Button signInButton = view.findViewById(R.id.sign_in);
        signInButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String phone = mPhoneView.getText().toString();
                String pwd = mPasswordView.getText().toString();

                if (mCallback != null) {
                    // mock login ok
                    mCallback.onLoginOk(true, new User(
                            UUID.randomUUID().toString(),
                            "Jack",
                            mPhoneView.getText().toString()
                    ));
                }
            }
        });

        return view;
    }

    public void setLoginCallback(LoginByPwdCallback callback) {
        this.mCallback = callback;
    }


    public interface LoginByPwdCallback {

        void onLoginOk(boolean needSmsVerify, User user);

    }
}

File: app/src/main/java/prototypez/github/io/sq/demo/processes/login/steps/LoginByPwdFragment.kt
package prototypez.github.io.sq.demo.processes.login.steps

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.jakewharton.rxbinding2.view.RxView
import kotlinx.android.synthetic.main.fragment_login_by_pwd.*
import prototypez.github.io.sq.demo.R
import prototypez.github.io.sq.demo.entity.User
import prototypez.github.io.sq.demo.processes.pwdReset.PwdResetActivity
import prototypez.github.io.sq.demo.processes.register.RegisterActivity
import java.util.*

class LoginByPwdFragment : Fragment() {

    lateinit var loginByPwdCallback: (needSmsVerify: Boolean, user: User) -> Unit
    lateinit var registerCallback: (User) -> Unit

    private val random = Random()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_login_by_pwd, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        sign_in.setOnClickListener {
            val phone = et_phone.text.toString()
            val pwd = et_password.text.toString()

            if (random.nextBoolean()) {
                // 50% chance to enter sms verify
                loginByPwdCallback(true, User(
                        UUID.randomUUID().toString(),
                        "Jack",
                        et_phone.text.toString()
                ))
            } else {
                // 50% chance login ok immediately
                loginByPwdCallback(false, User(
                        UUID.randomUUID().toString(),
                        "Jack",
                        et_phone.text.toString(),
                        "Shanghai",
                        29,
                        false
                ))
            }
        }

        // 嵌套流程：注册
        RxView.clicks(tv_register)
                .compose(RegisterActivity.register(this))
                .subscribe { user ->
                    registerCallback(user)
                }


        // 嵌套流程：忘记密码
        RxView.clicks(tv_reset_pwd)
                .compose(PwdResetActivity.pwdReset(this))
                .subscribe()
    }
}
